src/compiler/parser.ts

```diff
@@ -2739,49 +2739,52 @@ namespace ts {
-                    case SyntaxKind.ExclamationToken:
-                        return SyntaxKind.JSDocNonNullableType;
-                    case SyntaxKind.QuestionToken:
-                        return SyntaxKind.JSDocNullableType;
-                }
-            }
-        }
-
-        function parseArrayTypeOrHigher(): TypeNode {
-            let type = parseJSDocPostfixTypeOrHigher();
-            while (!scanner.hasPrecedingLineBreak() && parseOptional(SyntaxKind.OpenBracketToken)) {
-                if (isStartOfType()) {
-                    const node = <IndexedAccessTypeNode>createNode(SyntaxKind.IndexedAccessType, type.pos);
-                    node.objectType = type;
-                    node.indexType = parseType();
-                    parseExpected(SyntaxKind.CloseBracketToken);
-                    type = finishNode(node);
+        function parsePostfixTypeOrHigher(): TypeNode {
+            let kind: SyntaxKind | undefined;
+            let type = parseNonArrayType();
+            while (!scanner.hasPrecedingLineBreak() && (kind = getPostfixSyntaxKind(token()))) {
+                if (kind === SyntaxKind.OpenBracketToken) {
+                    parseExpected(SyntaxKind.OpenBracketToken);
+                    if (isStartOfType()) {
+                        const node = createNode(SyntaxKind.IndexedAccessType, type.pos) as IndexedAccessTypeNode;
+                        node.objectType = type;
+                        node.indexType = parseType();
+                        parseExpected(SyntaxKind.CloseBracketToken);
+                        type = finishNode(node);
+                    }
+                    else {
+                        const node = createNode(SyntaxKind.ArrayType, type.pos) as ArrayTypeNode;
+                        node.elementType = type;
+                        parseExpected(SyntaxKind.CloseBracketToken);
+                        type = finishNode(node);
+                    }
                 }
                 else {
-                    const node = <ArrayTypeNode>createNode(SyntaxKind.ArrayType, type.pos);
-                    node.elementType = type;
-                    parseExpected(SyntaxKind.CloseBracketToken);
-                    type = finishNode(node);
+                    nextToken();
+                    const postfix = createNode(kind, type.pos) as JSDocOptionalType | JSDocNonNullableType | JSDocNullableType;
+                    postfix.type = type;
+                    type = finishNode(postfix);
                 }
             }
             return type;
         }
 
+        function getPostfixSyntaxKind(tokenKind: SyntaxKind): SyntaxKind {
+            switch (tokenKind) {
+                case SyntaxKind.EqualsToken:
+                    // only parse postfix = inside jsdoc, because it's ambiguous elsewhere
+                    return contextFlags & NodeFlags.JSDoc ? SyntaxKind.JSDocOptionalType : 0;

```

Comment:
If you are trying to prevent boxing, there's a `SyntaxKind.Unknown`

Reply:
I was trying to prevent boxing **and**  `SyntaxKind.Unknown` at the same time. Actually, there's not much point, because `0` requires a falsy check later, and functions automatically union their return types these days. I might as well return false here.